### DISCUSSION TITLE
approaches to removing alpha channel (CLI)

### DISCUSSION THREAD

**Discussion by jrochkind** [libvips/libvips#3589] [Q&A] [answered]
I am using `vipsthumbnail` to take TIFF input, and resize it to a smaller JP2 output. 

The final destination is to pass as an argument to the `img2pdf` python tool, which can't handle alpha channels in input images (for JP2 at least).  While my images aren't intended to have alpha channels, occasionally one slips through that does. 

I have discovered I can use `vips flatten` to remove the alpha channel...

* This is a bit dangerous, because if I accidentally run it on an image that _doesn't have_ an alpha channel, it will still flatten one of the three existing channels, resulting in a two-channel image with colors corrupted. Is there anything i can do to flatten -only- an actual fourth alpha channel and no-op if it doesn't have one?

* It would be nice if I could do the flattening without having to invoke a separate CLI command -- especially because it will take another read/write from disk, taking time)
   *  It doens't appear there's any flatten option in the `vipsthumbnail` tool, unless I'm missing it. Would it make sense to add it as an option in the future?   Are there any other CLI commands available that could translate to a new format, resize, and flatten all in one execution?
   * I guess maybe I could _pipe_ the results of `vipsthumbnail` to `vips flatten` to at least avoid the extra disk read/write of the extra flatten... oh wait, I don't think vipsthumbnail can actually write to stdout (because it's designed for multi-file operations, makes sense).   Hmm, not sure if there's any other way to assemble CLI tools to do this... the danger of accidentally flattening a 3-channel file that shouldn't be flattened makes it somewhat more dangerous too. 

* Would these things be more feasible if I were using ruby library (or python) instead of trying to stick to CLI?

**Comment by jcupitt** [comment, marked as answer]
Hi @jrochkind,

In python it's just:

```python
if image.hasalpha():
    image = image.flatten()
```

Ruby is something like:

```ruby
image = image.flatten if image.has_alpha?
```

The libvips CLI is deliberately very basic. The idea was that anything fancy should be done in a proper scripting language anyway, so there's no point implementing a complete CLI environment. It's much better to make the library easy to bind to, and then pass the programming on to some other system.

libvips will flatten automatically if the destination format does not support alpha, but you're right, there's no way to force a flatten.

You can `vipsthumbnail` to a pipe with eg.:

```
$ vipsthumbnail k2.jpg -o .png | cat >x.png
```

Or even:

```
$ cat k2.jpg | vipsthumbnail stdin -o .png | cat >x.png
```

**Reply by jrochkind** [reply]
Thank you for quick response!

I guess I have tried to stick to CLI for too long!  For reasons, but it's probably time to bite the bullet. 

Meanwhile... I'm right that there's no way to have `vips flatten` only flatten an actual alpha, and not flatten 3 channels into 2 if I mistakenly run it on 2?

OK, so I think I need to first check how many channels it has. Then only if it has 4 (not sure what will happen with more than 4 or how that might be possible!), do the thing. 

Can `vips flatten` CLI read from stdin?  So I could do something like, I dunno....

    vips thumbnail k2.tiff -o .jp2 | vips flatten - output.jp2

**Reply by jcupitt** [reply]
CMYKA images (for example) will have five bands. I think I'd just use python.

Use the magic filename `stdin` to mean read-from-stdin. So:

```
vips thumbnail k2.tiff -o .ppm | vips flatten stdin output.jp2
```

**Reply by jrochkind** [reply]
Thanks!

We don't really use python here, but do use ruby... but I'm a bit scared to go over that hump right now to ruby SDK instead of CLI, I'm worried it will wind up having other unanticipated side effects with memory management and stuff (eg once a ruby allocates more RAM... it's total RAM allocations never go down again... but not sure if that will apply to C bindings like here... but when you are just shelling out to a CLI you _know_ it's entirley independent and can't mess up your ruby processes memory management or concurrency controls etc). 

But I hear you that our use cases have crossed the line to sophisticated enough that we really really ought to.

**Reply by jcupitt** [reply]
ruby-vips is very well tested -- it's the default image handler for rails now. You can make a little script that does your task, then shell out to that, and you get the same guarantees about memory release that you'd get with the CLI.

**Reply by jrochkind** [reply]
Very good point! Thank you for your advice!
